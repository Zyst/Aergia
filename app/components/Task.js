// @flow
/* eslint-disable react/no-unused-prop-types */
import React from "react";

type Props = {
  task: {
    name?: string,
    currentProgress?: number,
    totalPomodoros?: number,
    stopped?: number,
    active?: boolean
  },
  completeTask: (name: string) => void,
  removeTask: (name: string) => void,
  activateTask: (name: string) => void,
  deactivateTask: (name: string) => void
};

const Task = ({
  task,
  completeTask,
  removeTask,
  activateTask,
  deactivateTask
}: Props) => {
  const { name, stopped, buttons } = task.active
    ? {
        name: "task-active-name",
        stopped: "task-active-stopped pad-task u-pull-right",
        buttons: "pad-task u-pull-right"
      }
    : {
        name: "task-name",
        stopped: "task-stopped pad-task u-pull-right",
        buttons: "pad-task u-pull-right"
      };

  return (
    <div className="container task-bottom-pad">
      <span className={name}>
        {task.name}
      </span>

      <span className="u-pull-right">
        {task.currentProgress}/{task.totalPomodoros}
      </span>

      {task.stopped
        ? <span className={stopped}>
            Stopped {task.stopped} times
          </span>
        : ""}

      {task.active
        ? <span className={buttons}>
            <i
              onClick={() => completeTask(task.name || "")}
              role="button"
              tabIndex="0"
              className="fa fa-check success"
            />
            <i
              onClick={() => deactivateTask(task.name || "")}
              role="button"
              tabIndex="0"
              className="fa fa-stop danger"
            />
          </span>
        : <span className={buttons}>
            <i
              onClick={() => activateTask(task.name || "")}
              role="button"
              tabIndex="0"
              className="fa fa-chevron-up success"
            />
            <i
              onClick={() => removeTask(task.name || "")}
              role="button"
              tabIndex="0"
              className="fa fa-times danger"
            />
          </span>}
    </div>
  );
};

export default Task;
