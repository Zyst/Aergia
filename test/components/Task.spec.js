import { spy } from "sinon";
import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Task from "../../app/components/Task";

function setup() {
  const actions = {
    completeTask: spy(),
    removeTask: spy(),
    activateTask: spy(),
    deactivateTask: spy()
  };

  const activeMockTask = {
    name: "Active mock",
    currentProgress: 1,
    totalPomodoros: 3,
    stopped: 1,
    active: true
  };

  const inactiveMockTask = {
    name: "mock",
    currentProgress: 0,
    totalPomodoros: 1,
    stopped: 1,
    active: false
  };

  const activeComponent = shallow(<Task task={activeMockTask} {...actions} />);
  const inactiveComponent = shallow(
    <Task task={inactiveMockTask} {...actions} />
  );

  return {
    actions,
    activeMockTask,
    inactiveMockTask,
    activeComponent,
    inactiveComponent
  };
}

describe("Task component", () => {
  it("should be defined", () => {
    expect(Task).toBeDefined();
  });

  it("should match snapshot", () => {
    const { inactiveMockTask, activeMockTask, actions } = setup();

    const inactiveTree = renderer
      .create(<Task task={inactiveMockTask} {...actions} />)
      .toJSON();

    expect(inactiveTree).toMatchSnapshot("inactive");

    const activeTree = renderer
      .create(<Task task={activeMockTask} {...actions} />)
      .toJSON();

    expect(activeTree).toMatchSnapshot("active");

    const noStopTree = renderer.create(
      <Task
        task={{
          ...activeMockTask,
          stopped: 0
        }}
        {...actions}
      />
    );

    expect(noStopTree).toMatchSnapshot("can't stop won't stop");
  });

  it("should call complete task when clicked", () => {
    const { activeComponent, actions } = setup();
    const checkmark = activeComponent.find(".fa.fa-check.success").first();

    checkmark.simulate("click");
    expect(actions.completeTask.called).toBe(true);
  });

  it("should call deactivate task when clicked", () => {
    const { activeComponent, actions } = setup();
    const stop = activeComponent.find(".fa.fa-stop.danger").first();

    stop.simulate("click");
    expect(actions.deactivateTask.called).toBe(true);
  });

  it("should call remove task when clicked", () => {
    const { inactiveComponent, actions } = setup();
    const remove = inactiveComponent.find(".fa.fa-times.danger").first();

    remove.simulate("click");
    expect(actions.removeTask.called).toBe(true);
  });

  it("should call activate task when clicked", () => {
    const { inactiveComponent, actions } = setup();
    const activate = inactiveComponent
      .find(".fa.fa-chevron-up.success")
      .first();

    activate.simulate("click");
    expect(actions.activateTask.called).toBe(true);
  });
});
