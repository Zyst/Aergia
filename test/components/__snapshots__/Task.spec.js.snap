// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`active 1`] = `
<div
  className="container task-bottom-pad"
>
  <span
    className="task-active-name"
  >
    Active mock
  </span>
  <span
    className="u-pull-right"
  >
    1
    /
    3
  </span>
  <span
    className="task-active-stopped pad-task u-pull-right"
  >
    Stopped 
    1
     times
  </span>
  <span
    className="pad-task u-pull-right"
  >
    <i
      className="fa fa-check success"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
    <i
      className="fa fa-stop danger"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
  </span>
</div>
`;

exports[`can't stop won't stop 1`] = `
<div
  className="container task-bottom-pad"
>
  <span
    className="task-active-name"
  >
    Active mock
  </span>
  <span
    className="u-pull-right"
  >
    1
    /
    3
  </span>
  
  <span
    className="pad-task u-pull-right"
  >
    <i
      className="fa fa-check success"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
    <i
      className="fa fa-stop danger"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
  </span>
</div>
`;

exports[`inactive 1`] = `
<div
  className="container task-bottom-pad"
>
  <span
    className="task-name"
  >
    mock
  </span>
  <span
    className="u-pull-right"
  >
    0
    /
    1
  </span>
  <span
    className="task-stopped pad-task u-pull-right"
  >
    Stopped 
    1
     times
  </span>
  <span
    className="pad-task u-pull-right"
  >
    <i
      className="fa fa-chevron-up success"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
    <i
      className="fa fa-times danger"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    />
  </span>
</div>
`;
